### ISSUE TITLE
Allow `Matrix.upload` to accept non-square matrices

### ISSUE THREAD

**Issue by Yuriyzabegaev** [shwina/pyamgx#23]
Hello! 
Currently I'm working with sparse square matrix formed by PDE system. Matrix is singular as there are some zero rows, including last ones. It's not a lineal algebra issue, because RHS is also zero at these points. Still, I can't use `Matrix.upload` method. It checks CSR `indptr` size and decides, that matrix is non-square (but it is square). `indptr` gives incorrect matrix size, because last rows of matrix have no non-zero values. I made a change in that code and AMGX works correctly with such a matrix.
I suggest 2 things:

- not to check matrix shape at all
- give a warning if something is wrong with shape, such as at my case

Also in MPI case AMGX `Matrix.upload` function accepts non-square matrix. It is stated in AMGX documentation. So I think this check is not needed at all. I can make pull request for this feature if you decide, what option is preferable.

**Comment by shwina** [comment]
@Yuriyzabegaev thanks for reporting. This was definitely an oversight on my part. I think option (1) makes the most sense, i.e., let's not check the matrix shape at all.

>  I can make pull request for this feature if you decide, what option is preferable.

I would love if you submitted a pull request adding this feature!

**Comment by Yuriyzabegaev** [comment]
Should I change tests for rectangular matrices, so it will be possible to upload them?

**Comment by shwina** [comment]
> Should I change tests for rectangular matrices, so it will be possible to upload them?

Yes - thank you!

**Comment by qingshanchen** [comment]
But the real question is whether the underlying AMGX library can actually handle non-square matrices. If not, I would like to see the shape-checking feature to remain in pyamgx.

Regards,
Qingshan

On Jan 11, 2021, at 10:25 AM, Ashwin Srinath <notifications@github.com> wrote:

﻿

Closed #23<https://github.com/shwina/pyamgx/issues/23>.

**Comment by shwina** [comment]
@Yuriyzabegaev I think it makes sense to go with option (1). You're right, looking at the `indptr` isn't sufficient to

**Comment by shwina** [comment]
@Yuriyzabegaev after your latest changes, what's the behaviour o when we actually attempt to solve a matrix system with a rectangular matrix? Do we currently raise a nice error message? Or do we segfault?

**Comment by Yuriyzabegaev** [comment]
@shwina @qingshanchen thanks for the comments! I didn’t mean AMGX can solve non-square systems. What I said is that `Matrix.upload` method of AMGX can take rectangular chunk of square matrix when using distributed memory in MPI. This is stated in API reference: https://github.com/NVIDIA/AMGX/blob/main/doc/AMGX_Reference.pdf

![36325A01-085C-4981-8208-098E4A84EA4C](https://user-images.githubusercontent.com/24578236/104225924-18829b80-5458-11eb-9db9-b309b9506725.png)

Also it’s written there, that AMGX itself can’t work with non-square matrices. So passing it will lead to segfault. But it’s impossible to determine correct matrix shape only from its CSR structures in cases like mine. 
That seems consistent for me to keep things like they are now - to segfault or except internal AMGX error on invalid input, because PyAMGX is a wrapper - why should we add any logic above?
Nevertheless, I’m open for discussion!

**Comment by qingshanchen** [comment]
I see your point. But I don’t think that doing so will solve your problem. I am no expert on AMG, but I don’t think it can handle non-square matrix or square but singular matrices. The document you referenced, if you read carefully, the block matrix could be non-square because halo matrices are included. Therefore, even in distributed cases, AMGX is still expecting square matrices (it will internally trim the halo parts, it appears).

Regards,
Qingshan

On Jan 11, 2021, at 2:09 PM, Юрий <notifications@github.com> wrote:

﻿

@shwina<https://github.com/shwina> @qingshanchen<https://github.com/qingshanchen> thanks for the comments! I didn’t mean AMGX can solve non-square systems. What I said is that Matrix.upload method of AMGX can take rectangular chunk of square matrix when using distributed memory in MPI. This is stated in API reference: https://github.com/NVIDIA/AMGX/blob/main/doc/AMGX_Reference.pdf

[36325A01-085C-4981-8208-098E4A84EA4C]<https://user-images.githubusercontent.com/24578236/104225924-18829b80-5458-11eb-9db9-b309b9506725.png>

Also it’s written there, that AMGX itself can’t work with non-square matrices. So passing it will lead to segfault. But it’s impossible to determine correct matrix shape only from its CSR structures in cases like mine.
That seems consistent for me to keep things like they are now - to segfault or except internal AMGX error on invalid input, because PyAMGX is a wrapper - why should we add any logic above?
Nevertheless, I’m open for discussion!

**Comment by shwina** [comment]
@Yuriyzabegaev for an example matrix with zero rows, how do you compute the `indptr`, `indices` and `data`?

For example, the following matrix in CSR format has the attributes below:

```
1 2 3
1 2 4
0 0 0

indptr: [0, 3, 6, 6]
indices: [0, 1, 2, 0 1, 2]
data: [1, 2, 3, 1, 2, 4]
```

Note the repeated element `6` in `indptr` is used to indicate a zero row.

**Comment by Yuriyzabegaev** [comment]
@shwina seems like I confused `indices` with `indptr` and rows with columns. In your matrix shape can be computed correctly. The issue happens with matrix below:
```
import numpy as np
from scipy.sparse import csr_matrix

a = csr_matrix(np.array([[1, 2, 0],
                         [4, 5, 0],
                         [7, 8, 0]]))
nrows = len(a.indptr) - 1
ncols = a.indices.max() + 1
print(nrows, ncols)
```
It gives:
`3 2`

**Comment by shwina** [comment]
@Yuriyzabegaev @qingshanchen one solution here is to require a `shape` parameter in the `Matrix.upload` method.

Would that be reasonable? Is there any situation where you don't know (or can't easily compute) the shape of the matrix you wish to upload?

Note: `upload_csr` will not need change as we can get the shape from the CSR matrix.

**Comment by Yuriyzabegaev** [comment]
@shwina @qingshanchen only the number of rows is required for AMGX. You can always determine it from `indptr`, even for matrix with zero rows like above. Number of columns is required only to check if the matrix is square.
I can’t imagine a situation when determining matrix shape was hard. But it will change function interface. So maybe we should make `shape` parameter with default value `None`? It will check shape if it’s provided, otherwise it will just pass number of rows to AMGX.

**Comment by shwina** [comment]
To be clear, these are the changes I'm thinking need to be made:

* We should accept a `shape` parameter in the `Matrix.upload()` method. Rectangular shapes are allowed. If the shape is not provided, it is inferred from the data. One situation in which it cannot always be inferred correctly is when there are one or more columns of zeros.

* We should check that the matrix is square in the `Solver.solve()` method and raise early if it is not.

**Comment by Yuriyzabegaev** [comment]
@shwina I got your idea.
Ok, I’m working on it.

**Comment by shwina** [comment]
Closed by #27
